### PR TITLE
25 normalize data on gpu

### PR DESCRIPTION
**Pull request by sadamov**
## Summary
This PR introduces `on_after_batch_transfer` logic to normalize data on GPU instead of on CPU before transfer.

### Rationale
Normalization is faster on GPU than CPU. In the current code the data was normalized in the pytorch dataset class in the __get_item__ method potentially slowing down training; especially on systems with fewer CPU cores.

### Changes
- Introduces `on_after_batch_transfer` in the `ar_model.py` script
- Remove everything related to "standardization" from the remaining scripts
- Adapted the `create_parameter_weights.py` script to work with the new changes (not reloading standardized dataset)

### Testing
Both training and evaluation was successful. The training loss of 3.230 on the meps_example is identical to before the changes. The create_parameter_weights script was executed to successfully generate the stats.

### Not-In-Scope
The normalization stats and other static features will all become zarr archives in the future. There path defined in the data_config.yaml file.

### CODE REVIEW

**Comment by leifdenby** [comment]
This is looking good @sadamov! Can I suggest we merge #38 first, you merge that into your branch and then I do a review? That way we can ensure everything keeps working :)

**Review by joeloskarsson — CHANGES_REQUESTED** [review]
This is a lot cleaner and also good to move computations from CPU to GPU. There is a question mark about forcing standardization that should be double-checked. In the future we might want to handle standardization of forcing the same for all forcing variables, or have a more robust way to control this (i.e. which forcing should be standardized or not), but let's leave that to then.

File: neural_lam/models/ar_model.py [hunk 1]
```diff
@@ -197,6 +197,19 @@ def common_step(self, batch):
 
         return prediction, target_states, pred_std
 
+    def on_after_batch_transfer(self, batch, dataloader_idx):
+        """Normalize Batch data after transferring to the device."""
+        init_states, target_states, forcing_features = batch
+        init_states = (init_states - self.data_mean) / self.data_std
+        target_states = (target_states - self.data_mean) / self.data_std
+        forcing_features = (forcing_features - self.flux_mean) / self.flux_std
```
[the remark this answers is not in the document]
Comment (sadamov) [reply neural_lam/models/ar_model.py:202]: True, the forcings are now handled differently. I suggest to implement a new logic to handle forcings in #54. The use can define combined_vars that share statistics and also define vars that should not be normalized.

**Comment by sadamov** [comment]
I merged with the latest updates from main @leifdenby and commented on your suggestions @joeloskarsson. In the following I want to show that the output tensor was not affected by this change. As you suggested I stored the last batch for both gpu/cpu normalization with `deterministic=True` and a set seed: 

Pasted image

Then I compared the tensors like this:

Screenshot from 2024-06-07 21-06-37

So except for the forcings tensor which are handled differently the two approaches create identical output.

**Comment by joeloskarsson** [comment]
How do you think we should progress with this @sadamov ? If the forcing is handled differently in #54, would it make more sense to try to merge this after that? (I guess baking this change into #54 would just make it even bigger). Or should we merge this first so that #54 can build on it and be adapted to use it?

I would not be happy to merge this without fixing so also the forcing tensors match the previous implementation. But we could just do a quick fix for now so the standardization is only applied to the flux dimensions of the forcing tensor?

**Comment by sadamov** [comment]
I propose to merge #54 first and leave this open until then. We now know that `on_after_batch_transfer` works as expected and can shortly after implement this PR.

**Comment by joeloskarsson** [comment]
As we look at this after #66 we should think about more options for rescaling of the different variables. In #66 all variables (including state/forcing/static) are standardized. There are benefits to allowing for all three of:

1. Standardize to $\mu=0, \sigma=0$
2. Normalize to $[0,1]$
3. Perform no rescaling

for different variables. We do however need a way to specify what should be used for each variable, as well as computing the needed data statistics. Some of this might have to be done in mllam_data_prep, but at least the final computation should be in scope for this PR, as it is what the code has to do on the GPU.

**Comment by sadamov** [comment]
Please delete the PR-branch after merging, it's the last branch on mllam/neural_lam instead of fork.

**Comment by joeloskarsson** [comment]
@sadamov What should we do with this PR? The code here is really outdated, but in principle all of the logic is sound and we could update this to work on `main`. At the same time, with the datastores and zarr-based data-loading computing the standardization on CPU in the dataset class has not been any performance issue, so maybe this should not be a priority. One could argue that it adds complexity (harder to understand where standardization happens) for unnecessary performance improvement.

**Comment by sadamov** [comment]
Yes I agree, this was not an issue on most systems. I converted it to draft, so that we keep the solution but don't prioritize the implementation.

**Comment by sadamov** [comment]
superceded by #239
